src/workflow/default_executor.ts

```diff
@@ -15,18 +25,23 @@ export class DefaultWorkflowExecutor implements WorkflowExecutor {
       setContext: (key, value) => {
         context[key] = value;
       },
+      originalInput: input,
     };
 
-    let finalOutput: any;
+    let state: any = input;
+    let currentInput: any = input;
     for (const stepExecutor of engine.stepExecutors) {
       const step = stepExecutor.getStep();
       try {
-        const { skipped, output } = await stepExecutor.execute(input, executionBindings);
+        const { skipped, output } = await stepExecutor.execute(currentInput, executionBindings);
         if (skipped) {
           continue;
         }
         executionBindings.outputs[step.name] = output;
-        finalOutput = output;
+        state = output;
+        if (this.options.chainOutputs) {
+          currentInput = state;
```

Comment:
By replacing the value in `currentInput`, we lost access to the actual input to the step. Not that we need the input in the current implementation, but a slight alteration would improve the readability and cater to these needs in the future.

Something like this:

```
let curStepInput = input;
let prevStepOutput;
for (...) {
  if (chain) {
    curStepInput = prevStepOutput;
  }
  
  const curStepOutput = stepExecute(curStepInput);
  prevStepOutput = curStepOutput;
  // we can still make use of both the input and output of this step beyond this point
}
```